engine/classes/Elgg/AjaxApi/Service.php

```diff
@@ -0,0 +1,196 @@
+		$this->hooks = $hooks;
+		$this->msgs = $msgs;
+		$this->input = $input;
+	}
+
+	/**
+	 * Did the request come from the elgg/ajax module?
+	 *
+	 * @return bool
+	 */
+	public function isAjax2Request() {
+		$version = _elgg_services()->request->headers->get('X-Elgg');
+		return ($version === '2');
+	}
+
+	/**
+	 * Can this service respond to the request?
+	 *
+	 * @return bool
+	 */
+	public function canRespond() {
+		return !$this->response_sent && $this->isAjax2Request();
+	}
+
+	/**
+	 * Attempt to JSON decode the given string
+	 *
+	 * @param mixed $string Output string
+	 * @return mixed
+	 */
+	public function decodeJson($string) {
+		if (!is_string($string)) {
+			return $string;
+		}
+		$object = json_decode($string);
+		return ($object === null) ? $string : $object;
+	}
+
+	/**
+	 * Send a JSON HTTP response with the given output
+	 *
+	 * @param mixed  $output     Output from a page/action handler
+	 * @param string $hook_type  The hook type. If given, the response will be filtered by hook
+	 * @param bool   $try_decode Try to convert a JSON string back to an abject
+	 * @return void
+	 */
+	public function respondFromOutput($output, $hook_type = '', $try_decode = true) {
+		if ($try_decode) {
+			$output = $this->decodeJson($output);
+		}
+
+		$api_response = new ApiResponse();
+		$api_response->setData($output);
+		$api_response = $this->filterApiResponse($api_response, $hook_type);
+		$response = $this->buildHttpResponse($api_response);
+
+		$this->response_sent = true;
+		$response->send();
+	}
+
+	/**
+	 * Send a JSON HTTP response based on the given API response
+	 *
+	 * @param AjaxApiResponse $api_response API response
+	 * @param string          $hook_type    The hook type. If given, the response will be filtered by hook
+	 * @return void
+	 */
+	public function respondFromApiResponse(AjaxApiResponse $api_response, $hook_type = '') {
+		$api_response = $this->filterApiResponse($api_response, $hook_type);
+		$response = $this->buildHttpResponse($api_response);
+
+		$this->response_sent = true;
+		$response->send();
+	}
+
+	/**
+	 * Send a JSON HTTP 400 response
+	 *
+	 * @param string $msg The error message (not displayed to the user)
+	 * @return void
+	 */
+	public function respondWithError($msg) {
+		$response = new JsonResponse(['error' => $msg], 400);
+
+		$this->response_sent = true;
+		$response->send();
+	}
+
+	/**
+	 * Filter an AjaxApiResponse through a plugin hook
+	 *
+	 * @param AjaxApiResponse $api_response The API Response
+	 * @param string          $hook_type    The hook type. If given, the response will be filtered by hook
+	 *
+	 * @return AjaxApiResponse
+	 */
+	private function filterApiResponse(AjaxApiResponse $api_response, $hook_type = '') {
+		$api_response->setTtl($this->input->get('response_ttl', 0, false));
+
+		if ($hook_type) {
+			$hook = AjaxApiResponse::RESPONSE_HOOK;
+			$api_response = $this->hooks->trigger($hook, $hook_type, null, $api_response);
```

Comment:
Here are the new hooks, all varying by types, so they're very granular. Easiest to see the types in the AMD module.